### DISCUSSION TITLE
Feature Wishlist

### DISCUSSION THREAD

**Discussion by VisenDev** [neurocyte/flow#62] [Ideas]
A list of features and suggestions

* Pressing a prefix key in vim/helix mode should show a menu of possible completions. Example (from helix):
<img width="1000" alt="Screenshot 2024-10-18 at 2 16 24 PM" src="https://github.com/user-attachments/assets/d77e1d06-7af4-49db-b817-ad9c1934457d">

* A Command to describe the currently active keybindings in any buffer. [Nyxt Browser](https://nyxt-browser.com/documentation#the-nyxt-help-system) has this feature and I found it very helpful when getting started (Similar to what `C-h b` does in emacs)
   
* Enable widely used shortcuts for exiting insert mode by default (such as `jk`). See [this](https://github.com/helix-editor/helix/discussions/11710) for more related discussion

* Implement emacs keybindings mode
  * Also, It would be really cool if it was possible to use emacs style keybindings for navigation inside of vim insert mode (similar to how compatible flow keybindings are enabled inside insert mode)
  
* Don't rely on plugins for things which should be features
   * Vim and Neovim particularly irritate me regarding this. The default state of an editor should be acceptable for prolonged use, with detailed configuration being *optional*. In Vim and Neovim, you are more or less required to deal with installing dozens of plugins and extensive configuration before you get a usable editor. The Helix editor is a great example of an editor that does not do this and has very usable defaults.
   
* Change the `open file` interface to use the "Command Palette UI" instead of only showing one file at a time at the bottom of the screen. The Command Palette is really nice because it lets you do fuzzy finding to find commands. It would be really nice if opening a file used this same interface

* Some sort of REPL integration, see the related helix issue https://github.com/helix-editor/helix/issues/2806. The basic idea is to have commands to send certain selected parts of a file to a REPL in another pane, so that code can be run and interacted with while actively writing it. 

* Add project level search and replace functionality (rename symbols across an entire codebase)

**Comment by neurocyte** [comment]
> * Pressing a prefix key in vim/helix mode should show a menu of possible completions.

I'm not sure this better than fuzzy search in the command palette.

> * A Command to describe the currently active keybindings in any buffer.

The command palette already lists all commands available in the current context and their keybindings.

> * Enable widely used shortcuts for exiting insert mode by default (such as `jk`). See [this](https://github.com/helix-editor/helix/discussions/11710) for more related discussion

I don't think this is possible without add some (perhaps minimal) lag. Avoiding lag is a fairly core priority for flow.

> * Implement emacs keybindings mode

This would be great.

> * Don't rely on plugins for things which should be features

I agree fully with this. Flow aims to be full featured without installing plugins or even configuring anything. It should be useful to most user right ouf of the box.

> * Change the `open file` interface to use the "Command Palette UI" instead of only showing one file at a time at the bottom of the screen. The Command Palette is really nice because it lets you do fuzzy finding to find commands. It would be really nice if opening a file used this same interface

You can use the `Open recent file` command for this. It has fuzzy search and sorts files by most recently used. That being said, the `open file` command will be extended eventually with a file browsing UI of some sort. That UI *might* include fuzzy filtering.

> * Some sort of REPL integration

This should be pretty trivial to add with a plugin, once we have plugin support. It could also be added to flow directly, but I think it is likely to be repl/language specific.

**Comment by VarLad** [comment]
May I add, a way to have panes in general? Viewing multiple files side by side would be quite nice.
Even better would be a way to have some kind of client-server model, similar to Kakoune and NeoVim. This would allow one to start Flow with `flow --server server_name`, and then one could open another terminal and type `flow -c server_name ~/hello.zig` and it'll open in the same instance of Flow. This feature would enable people to integrate Flow with their favorite terminal multiplexer.

**Comment by neurocyte** [comment]
> Even better would be a way to have some kind of client-server model, similar to Kakoune and NeoVim. This would allow one to start Flow with `flow --server server_name`, and then one could open another terminal and type `flow -c server_name ~/hello.zig` and it'll open in the same instance of Flow. This feature would enable people to integrate Flow with their favorite terminal multiplexer.

Remote control (or client/server as you called it) is definitely planned. Mutilple tty support too.

**Comment by VarLad** [comment]
> The basic idea is to have commands to send certain selected parts of a file to a REPL in another pane, so that code can be run and interacted with while actively writing it.

Regarding this, @neurocyte is it possible to:
1. define custom commands (in the command palette + keybinding)
2. Select the entire line where the cursor is on, or multiple lines starting from the line cursor is on, from the open file.
3. A way to combine 1 and 2. Running custom commands on the selected text.

This alone would let people write integration with their favorite terminal multiplexers, as the selected line could be piped to a pane opened in the multiplexer where a REPL is running. At least this shouldn't require plugin functionality as of now.

**Comment by neurocyte** [comment]
> Regarding this, @neurocyte is it possible to:
> 
> 1. define custom commands (in the command palette + keybinding)

What makes a command "custom"? You can add any commands you need, yes. You don't need to add a command for the functionality you describe though.

> 2. Select the entire line where the cursor is on, or multiple lines starting from the line cursor is on, from the open file.
> 3. A way to combine 1 and 2. Running custom commands on the selected text.

You are describing the `filter` command here. See the the sort keybinding for an example of how it's used: https://github.com/neurocyte/flow/blob/a9058ef949695bf6329666033c3ed1b8bbb1a8df/src/tui/mode/input/flow.zig#L173

**Comment by neurocyte** [comment]
I will convert this to a discussion. Please create individual issues for each item that becomes actional in the descussion.

**Comment by VisenDev** [comment]
relevant items converted into issues
